src/core/expression/qgsexpressionnodeimpl.cpp

```diff
@@ -1793,6 +1793,128 @@ QList<const QgsExpressionNode *> QgsExpressionNodeInOperator::nodes() const
   return lst;
 }
 
+
+QgsExpressionNodeBetweenOperator::~QgsExpressionNodeBetweenOperator()
+{
+  delete mNode;
+  delete mLowerBound;
+  delete mHigherBound;
+}
+
+QgsExpressionNode::NodeType QgsExpressionNodeBetweenOperator::nodeType() const
+{
+  return ntBetweenOperator;
+}
+
+bool QgsExpressionNodeBetweenOperator::prepareNode( QgsExpression *parent, const QgsExpressionContext *context )
+{
+  bool res = mNode->prepare( parent, context );
+  res = res && mLowerBound->prepare( parent, context );
+  res = res && mHigherBound->prepare( parent, context );
+  return res;
+}
+
+QVariant QgsExpressionNodeBetweenOperator::evalNode( QgsExpression *parent, const QgsExpressionContext *context )
+{
+  QgsExpressionNodeBinaryOperator lowBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boGE, mNode->clone(), mLowerBound->clone() };
+  QgsExpressionNodeBinaryOperator highBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boLE, mNode->clone(), mHigherBound->clone() };
```

Comment:
doesn't this approach mean we are evaluating mNode twice?

Reply:
I think you could avoid this by evaluating mNode upfront, and then using a static value node in lowBound/highBound instead.

Reply:
I refactored the evaluation code trying to optimize but also to handle NULLs when it is possible to return a boolean, it is a bit more complex now but I made sure that all code paths are covered by tests.